### ISSUE TITLE
[sklearn#34502] Consider switching to pyrefly for typechecking

### ISSUE THREAD

**Issue by godofecht** [flooooooooooow/flow#319]
**Mirrored from [scikit-learn#34502](https://github.com/scikit-learn/scikit-learn/issues/34502)**

Original title: Consider switching to pyrefly for typechecking
Original labels: 
Created: 2026-07-17T19:34:27Z

---

I'd like to propose adding [Pyrefly](https://pyrefly.org) for typechecking - it could either be added alongside mypy as a non-blocking CI check or migrated entirely. Happy to open the PR, checking here first to see if there's interest?

For transparency I work on the Pyrefly team, I've been looking into the performance of mypy vs Pyrefly on real repos and when I tried out sklearn the results looked promising enough that I thought I'd share :)

I did some initial investigation, running locally mypy 2.3.0 vs Pyrefly 1.1.1 (checking `sklearn/`) which shows Pyrefly is ~29x faster:

  | Checker   | Run        |
  |-----------|------------|
  | mypy      | **18.4 s** |
  | Pyrefly   | **0.6 s**  |

## How to make the switch:

  1. **Add the dependency** to the test/lint requirements: `pyrefly>=1.1.1`.
  2. **Add `[tool.pyrefly]` config** mirroring the existing `[tool.mypy]` config.
     ```toml
     [tool.pyrefly]
     project-includes = ["sklearn"]                        # matches `mypy sklearn/` in build_tools/linting.sh
     search-path = ["."]
     project-excludes = ["**/sklearn/externals*"]          # mirrors mypy's exclude
     preset = "legacy"                                      # matches mypy's non-strict defaults
     ignore-missing-imports = ["joblib.*", "sklearn.externals.*", "*"]  # mirrors ignore_missing_imports + follow_imports=skip
     ```
  3. **resolve/suppress error code differences** There are some differences between the errors pyrefly catches and what mypy does (79 new errors from what I can see running locally, details below), so we could either ignore them  (`# pyrefly: ignore [code]`) for now or see if there are genuine issues to address. If we opt for removing mypy we could also remove ~80 out of 82 mypy ignores that pyrefly doesn't complain about, so the total ignores would stay roughly the same.
  4. **Setup in CI** in .github/workflows/lint.yml (`build_tools/linting.sh`), mirroring the mypy
     step with a run pyrefly check step.

## Error differences - Mypy vs Pyrefly

  When I compared pyrefly and mypy output, I found 79 new errors from Pyrefly that aren't flagged by Mypy. Some might be real typing issues that should be resolved, but I'm not super familiar with the sklearn codebase so would require some deeper investigation. Here's a summary of the differences:

  | Error code | Count | Reason for difference | Suggested approach |
  |---|---|---|---|
  | `bad-override` | 72 | Estimator subclasses override base `fit`/`transform`/callback methods with looser signatures than Pyrefly's LSP-override check allows — 54 in `sklearn/` (mostly `cluster/*`), 18 in tests. Pyright/mypy are more permissive here. | Suppress - looks mostly benign; could align a few signatures over time |
  | `missing-attribute` | 3 | `pytest` narrowed to `None` via the optional-import guard in `estimator_checks.py` (`.mark`/`.param`) | Suppress |
  | `bad-argument-type` | 1 | `_openml.py:716` | **Review** and consider fixing |
  | `no-matching-overl

...(truncated)

---

## Flow context

This issue is mirrored from scikit-learn to explore whether Flow encounters the same problem or can approach it differently. Add analysis below.

**Comment by godofecht** [comment]
Moved to https://github.com/godofecht/flow-scikit/issues/104 in the flow-scikit repository. Further library-specific triage and implementation will happen there.
